firmware/gameport-adapter/ThrustmasterAttackThrottle.h

```diff
@@ -0,0 +1,370 @@
+
+    // I'm not sure if this optimization actually does anything on AVR.
+    // My thought is locality of reference may help to get the core loop
+    // to be more performant. This is memcpy'ed to packet at the end when timing
+    // doesn't matter.
+    uint8_t localPacket[packetSize];
+
+    // Halt interrupts only during the timing-sensitve part of the code
+    //
+    // This code is *very* timing-sensitive. There was a time where it was
+    // working reliably until I changed an unrelated part of the code. Not
+    // sure if that impacted inlining or code layout or something but it was
+    // enough to break the timing.
+    // I think this version is reliable. It has been in practice
+    // but I wouldn't rule out something else odd like this happening again.
+    const InterruptStopper interruptStopper;
+    {
+      trigger();
+
+      // the first clock has a 200us timeout, every following clock
+      // has a much shorter timeout. I've seen the gap between the
+      // trigger and the first clock be about 100us. So 200us should
+      // be plenty of margin. Having a long timeout on the first bit
+      // makes this simpler than having extra wait calls to account
+      // for this gap.
+      int timeout = 200;
+
+      // each byte is shifted into this variable
+      uint8_t shiftReg;
+
+      // current location in filling in the packet
+      uint8_t *buffer = localPacket;
+
+      for (uint8_t byte = 0; byte < packetSize; ++byte) {
+        shiftReg = 0;
+
+        // check for the start bit, it should always be 1
+        if (!m_clock.wait(Edge::falling, timeout)) {
+          log("timeout clk");
+          return false;
+        }
+
+        if (!m_data.get()) {
+          // bad start bit
+          return false;
+        }
+
+        // after the first bit in the packet, adjust the timeout.
+        // A clock pulse is 12us. so this should be enough although
+        // I don't know how precise the timeout is.
+        timeout = 24;
+
+        // the data portion of the byte
+        for (uint8_t bit = 0; bit < 8; ++bit) {
+
+          if (!m_clock.wait(Edge::falling, 200)) {
+            return false;
+          }
+
+          // shift in from the left
+          shiftReg = shiftReg >> 1 | m_data.get() << 7;
+        }
+
+        if (!m_clock.wait(Edge::falling, 200)) {
+          return false;
+        }
+
+        if (m_data.get()) {
+          // bad stop bit
+          log("bad stop bit");
+          return false;
+        }
+        *buffer++ = shiftReg;
+      }
+
+      // A simple check that the packet is complete and there is no more data
+      // A more comprehensive check would require some extra delays and checking
+      // the clock pin as well
+      //
+      // I'd like this to be a rising edge check instead of just true
+      // but couldn't get it to reliably detect both the data going low
+      // then high fast enough on the Arduino.
+      if (!m_data.wait(true, 200)) {
```

Comment:
I left as-is. I've had reliable packet reads in the latest code so far. I believe rather than watching the data line, the code probably could check one of the other pins to be sure it is really done. There was a pin that acted like an enable pin, so it could watch it go to disabled. When looking at the trace I ignored this pin as it wasn't needed for decoding the data so I don't recall what pin it was.